### DISCUSSION TITLE
Cross Compilation Between WiringPi Library and F'

### DISCUSSION THREAD

**Discussion by OwenNieuwenhuizen** [nasa/fprime#2671] [Q&A] [answered]
Hi everyone, 
I made a component to control a 2 axis gimbal connected to a pi cam. My group and I are running a mission using a Raspberry Pi 4b and I am trying to use the WiringPi library to send PWMs to the micro servos that control the direction of the gimbal. I got the component to build, but my deployment doesn't seem to like the functions that I am using from the WiringPi library. I believe that it is a cross compilation issue. Does this sound right/how would you recommend I go about fixing this problem?
Thx.

**Comment by LeStarch** [comment, marked as answer]
Do you have access to the wiring-pi library?  Is it an archive (`.a`) library, a shared-object (`.so`)?  Additionally, what cross-compiler are you using?

There are several solutions that you might try to resolve the issue:

1. Run a toolchain with wiring-pi installed.  This works best when a vendor (or raspberry pi) has given you a per-configured toolchain.
2. Create a sysroot from the RPI 4 filesystem and/or the wiring pi library.  Then set the `--sysroot` flag when building.  This works best when your cross-compiler works *except for missing libraries like wiring pi*. 
3. Build your own tool chain. This works best when you need to change libc/libc++ versions and add libraries.  Usually you do this if the cross-compile does not work correctly.
4. Build code on the RPI4.  This only works with an `aarch64` operating system installed on the PI.  This is why F´  does not baseline this approach: PI1, PI2, and PI3 often ran 32-bit OS not the 64bit version.

If you provide more information we can try and help select one of those options.

**Reply by OwenNieuwenhuizen** [reply]
I have been trying to use gcc-arm-linux-gnueabihf and its toolchain, but it hasn't been working so far. I am trying to avoid building directly on the Pi due to the size of fprime and the slow compilation speed. For reference though, I have implemented a component which has generated and built. This is what it looks like.

(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/Components# cd Gimbal
(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/Components/Gimbal# ls
CMakeLists.txt  Gimbal.cpp  Gimbal.fpp  Gimbal.hpp  Gimbal.template.cpp  Gimbal.template.hpp  WiringPi  docs

On the deployment side, I have not been able to get it to build which is where I am running into the issue. This is the buildout

(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/GDep# ls
CMakeLists.txt  README.md  WiringPi-master                    logs        settings.ini
Main.cpp        Top        build-fprime-automatic-native-old  master.zip

This is the error I am getting when I try to build.

[100%] Linking CXX executable ../bin/Linux/GDep
/usr/bin/ld: ../lib/Linux/libComponents_Gimbal.a(Gimbal.cpp.o): in function `Components::Gimbal::Gimbal(char const*)':
Gimbal.cpp:(.text+0x75): undefined reference to `wiringPiSetupGpio'
/usr/bin/ld: Gimbal.cpp:(.text+0x84): undefined reference to `pinMode'
/usr/bin/ld: Gimbal.cpp:(.text+0x93): undefined reference to `pinMode'
/usr/bin/ld: Gimbal.cpp:(.text+0x9d): undefined reference to `pwmSetRange'
/usr/bin/ld: ../lib/Linux/libComponents_Gimbal.a(Gimbal.cpp.o): in function `Components::Gimbal::run_handler(int, unsigned int)':
Gimbal.cpp:(.text+0x1f9): undefined reference to `pwmWrite'
/usr/bin/ld: Gimbal.cpp:(.text+0x244): undefined reference to `pwmWrite'
/usr/bin/ld: ../lib/Linux/libComponents_Gimbal.a(Gimbal.cpp.o): in function `Components::Gimbal::SET_SERVO1_POSITION_cmdHandler(unsigned int, unsigned int, unsigned char)':
Gimbal.cpp:(.text+0x302): undefined reference to `pwmWrite'
/usr/bin/ld: ../lib/Linux/libComponents_Gimbal.a(Gimbal.cpp.o): in function `Components::Gimbal::SET_SERVO2_POSITION_cmdHandler(unsigned int, unsigned int, unsigned char)':
Gimbal.cpp:(.text+0x478): undefined reference to `pwmWrite'
collect2: error: ld returned 1 exit status
gmake[3]: *** [GDep/CMakeFiles/GDep.dir/build.make:216: bin/Linux/GDep] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:5288: GDep/CMakeFiles/GDep.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:5295: GDep/CMakeFiles/GDep.dir/rule] Error 2
gmake: *** [Makefile:1339: GDep] Error 2
[ERROR] CMake erred with return code 2

Do you recognize what could be the issue here?

**Reply by LeStarch** [reply]
Those are linker issues.  The linker is failing to find the needed library.  How do you build WiringPi and what is within the WiringPi folder under Gimbal?

**Reply by OwenNieuwenhuizen** [reply]
When I did this I cloned the WiringPi git repo. I am a bit unsure of the structure though as I am new to this, but this is what is inside of the WiringPi directory

(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/Components/Gimbal/WiringPi# ls
COPYING.LESSER  People     VERSION  debian           devLib    gpio        pins       version.h  wiringPiD
INSTALL         README.md  build    debian-template  examples  newVersion  reinstall  wiringPi

**Reply by LeStarch** [reply]
Ok, the quickest way to link against that would be to run their build process.  Then find the library that it built (probably a `.a` or a `.so`) and add it to your mod-deps list.  In Gimbal's `CMakeList.txt` add:

```
set(MOD_DEPS
    "${CMAKE_CURRENT_LIST_DIR}/WiringPi/.../.../
)
```

Where `.../...` is the path to the library that is built.

Note: you'll also need to install the built Debian package on the target RPI 4.

**Reply by OwenNieuwenhuizen** [reply]
I made the change in the CMakeLists.txt file in my Gimbal component and now I am running into a build error for the component unfortunately.

-- Configuring done (111.5s)
CMake Error at fprime/cmake/target/build.cmake:83 (add_dependencies):
-- Generating done (20.5s)
  The dependency target
  "_mnt_c_Users_owenn_Command_Components_Gimbal_usr_lib_libwiringPi.so" of
  target "Components_Gimbal" does not exist.
Call Stack (most recent call first):
  fprime/cmake/target/build.cmake:141 (build_setup_build_module)
  fprime/cmake/target/target.cmake:116 (build_add_module_target)
  fprime/cmake/target/target.cmake:116 (cmake_language)
  fprime/cmake/target/target.cmake:155 (setup_single_target)
  fprime/cmake/module.cmake:48 (setup_module_targets)
  fprime/cmake/module.cmake:88 (generate_base_module_properties)
  fprime/cmake/API.cmake:228 (generate_library)
  Components/Gimbal/CMakeLists.txt:35 (register_fprime_module)


CMake Generate step failed.  Build files cannot be regenerated correctly.
gmake: *** [Makefile:1475: cmake_check_build_system] Error 1
[ERROR] CMake erred with return code 2
(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/Components/Gimbal# ls /usr/lib/libwiringPi.*
/usr/lib/libwiringPi.so
(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/Components/Gimbal

I think that the issue is with the CMakeLists.txt file though, but I don't see the issue.

(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/Components/Gimbal# cat CMakeLists.txt
####
# F prime CMakeLists.txt:
#
# SOURCE_FILES: combined list of source and autocoding files
# MOD_DEPS: (optional) module dependencies
# UT_SOURCE_FILES: list of source files for unit tests
#
####
set(SOURCE_FILES
  "${CMAKE_CURRENT_LIST_DIR}/Gimbal.fpp"
  "${CMAKE_CURRENT_LIST_DIR}/Gimbal.cpp"
)

# Uncomment and add any modules that this component depends on, else
# they might not be available when cmake tries to build this component.

# set(MOD_DEPS
#     Add your dependencies here
# )

# Add include directories for WiringPi
include_directories(/usr/local/include)

# Create the component library
add_library(Gimbal ${SOURCE_FILES})

set(MOD_DEPS
    "${CMAKE_CURRENT_LIST_DIR}/usr/lib/libwiringPi.so"
)

# Add WiringPi library to the target
target_link_libraries(Gimbal wiringPi)

# Register the F prime module
register_fprime_module()
(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/Components/Gimbal#

**Reply by OwenNieuwenhuizen** [reply]
I edited that and then build the component and then the deployment as well. Thank you so much for your help!

**Reply by OwenNieuwenhuizen** [reply]
Unfortunately, while both the component and the deployment build now, when I try to run fprime-gds --ip-client, it gives me this:

(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/GDep# fprime-gds --ip-client
[INFO] Autodetected artifacts root '/mnt/c/Users/owenn/Command/GDep/build-artifacts' from project settings.ini file.
[ERROR] /mnt/c/Users/owenn/Command/GDep/build-artifacts/Linux does not exist. Make sure to build.

Am I deploying this wrong?

**Reply by LeStarch** [reply]
The gds cannot run cross-compiled code locally.  Nor can it deploy for you. Thus you need to:

fprime-gds --ip-client --dictionary build-artifacts/raspberrypi/deployment/deploymentAppAi.xml

Then scp your code to the PI and run there.  This is explained in the cross compiling tutorial.

**Reply by LeStarch** [reply]
Also make sure to install wiring pi on the pi or it will fail to find the library

**Reply by LeStarch** [reply]
Good work getting a build.

**Reply by OwenNieuwenhuizen** [reply]
Thank you! Logically that makes a lot of sense, but when I ran the command, it gave me this error:

(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/GDep# fprime-gds --ip-client --dictionary build-artifacts/raspberrypi/deployment/deploymentAppAi.xml
[ERROR] Failed to parse arguments: Dictionary file build-artifacts/raspberrypi/deployment/deploymentAppAi.xml does not exist
usage: fprime-gds [-h] [-l LOGS] [--log-directly] [--log-to-stdout] [-d DEPLOYMENT] [--dictionary DICTIONARY] [--packet-spec PACKET_SPEC]
                  [-g {none,html}] [--gui-addr GUI_ADDR] [--gui-port GUI_PORT] [-n] [--app APP] [--zmq] [--zmq-server]
                  [--zmq-transport serverInUrl serverOutUrl] [--tts-port TTS_PORT] [--tts-addr TTS_ADDR]
                  [--file-storage-directory FILES_STORAGE_DIRECTORY] [--remote-sequence-directory REMOTE_SEQUENCE_DIRECTORY]
                  [--ip-address ADDRESS] [--ip-port PORT] [--ip-client] [--uart-device DEVICE] [--uart-baud BAUD]
                  [--comm-adapter {none,ip,uart}] [--comm-checksum-type {fixed,crc32}] [--output-unframed-data [OUTPUT_UNFRAMED_DATA]]

Run F prime deployment and GDS

options:
  -h, --help            show this help message and exit
  -l LOGS, --logs LOGS  Logging directory. Created if non-existent. [default: /mnt/c/Users/owenn/Command/GDep/logs]
  --log-directly        Logging directory is used directly, no extra dated directories created.
  --log-to-stdout       Log to standard out along with log output files
  -d DEPLOYMENT, --deployment DEPLOYMENT
                        Deployment installation/build output directory. [default: install_dest field in settings.ini]
  --dictionary DICTIONARY
                        Path to dictionary. Overrides automatic dictionary detection.
  --packet-spec PACKET_SPEC
                        Path to packet specification.
  -g {none,html}, --gui {none,html}
                        Set the desired GUI system for running the deployment. [default: html]
  --gui-addr GUI_ADDR   Set the GUI server address [default: 127.0.0.1]
  --gui-port GUI_PORT   Set the GUI server address [default: 5000]
  -n, --no-app          Do not run deployment binary. Overrides --app.
  --app APP             Path to app to run. Overrides automatic app detection.
  --zmq                 Switch to using the ZMQ transportation layer
  --zmq-server          Sets the ZMQ connection to be a server. Default: false (client)
  --zmq-transport serverInUrl serverOutUrl
                        Pair of URls used with --zmq to setup ZeroMQ transportation [default: ['ipc:///tmp/fprime-server-in',
                        'ipc:///tmp/fprime-server-out']]
  --tts-port TTS_PORT   Set the threaded TCP socket server port [default: 50050]
  --tts-addr TTS_ADDR   Set the threaded TCP socket server address [default: 0.0.0.0]
  --file-storage-directory FILES_STORAGE_DIRECTORY
                        Directory to store uplink and downlink files. Default: /tmp/root
  --remote-sequence-directory REMOTE_SEQUENCE_DIRECTORY
                        Directory to save command sequence binaries, on the remote FSW. Default: /seq
  --ip-address ADDRESS  Address of the IP adapter server. Default: 0.0.0.0
  --ip-port PORT        Port of the IP adapter server. Default: 50000
  --ip-client           Run the IP adapter as the client (connects to FSW running TcpServer)
  --uart-device DEVICE  UART device representing the FSW. Default: /dev/ttyACM0
  --uart-baud BAUD      Baud rate of the serial device. Default: 9600
  --comm-adapter {none,ip,uart}
                        Adapter for communicating to flight deployment. [default: ip]
  --comm-checksum-type {fixed,crc32}
                        Setup the checksum algorithm. [default: crc32]
  --output-unframed-data [OUTPUT_UNFRAMED_DATA]
                        Log unframed data to supplied file relative to log directory. Use '-' for standard out.
(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/GDep#

It seems like it is not recognizing that file path, though I am not quite sure why.

**Reply by LeStarch** [reply]
Right, you have to adjust it for your build....I don't know what your deployment name is and I am typing from memory.

Try doing a "find build-artifacts -name *.xml" to get the name of the dictionary

**Reply by LeStarch** [reply]
Sometimes it is hard for me to distinguish command that are "just copy and paste" and those that need adjustments.  If you have thoughts on that, I'd appreciate it as it befuddles my documentation too.

**Reply by LeStarch** [reply]
Also, I noticed I didn't at the -n flag.  You need that too.

**Reply by OwenNieuwenhuizen** [reply]
I tend to look specifically look for the command lines in order to parse through when new commands were implemented, but the lack of formatting and color definitely causes me issues as well when I copy and paste.

**Reply by OwenNieuwenhuizen** [reply]
What would the -n flag be part of?

**Reply by LeStarch** [reply]
The -n flag needs to be added to the fprime-gds call.

fprime-gds -n --dictionary ....

This means "run without launching an app (I'll launch it myself) but the app I do launch uses the following dictionary"

**Reply by OwenNieuwenhuizen** [reply]
(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/GDep# find build-artifacts -name *.xml
find: ‘build-artifacts’: No such file or directory
(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/GDep# find build-artifacts -name GDep.xml
find: ‘build-artifacts’: No such file or directory

That definitely makes sense on how to run it front, but their doesn't seem to be build-artifacts unless I am reading this very wrong, which is possible.

**Reply by LeStarch** [reply]
That directory is probably up a level.  Try running it in the Command directory.  Also put " around *.xml

E.g. "*.xml" as sometimes the shell will expand that, and you want find to handle it instead

**Reply by OwenNieuwenhuizen** [reply]
Thank you so much! That did make it run, though how do I make it launch?

**Reply by LeStarch** [reply]
Now you need to copy the binary from the "bin" folder under build-artifacts to the raspberry pi and then run it.  Give it the flag -a 0.0.0.0 and -p 50000

**Reply by OwenNieuwenhuizen** [reply]
Oh okay I think that makes sense. So would I just ssh into the pi and then do that.

**Reply by LeStarch** [reply]
scp will perform the copy. Ssh will allow you to run the binary

**Reply by OwenNieuwenhuizen** [reply]
Okay thank you so much! You saved me a lot of time and headache!

**Reply by LeStarch** [reply]
Glad to hear it!

**Reply by thomas-bc** [reply]
@OwenNieuwenhuizen also note that we have some material is explaining this process of crosscompilation/running on a remote raspberrypi in the [LedBlinker Tutorial](https://fprime-community.github.io/fprime-workshop-led-blinker/)

**Reply by OwenNieuwenhuizen** [reply]
I started trying to do the crosscompilation/running bit of that tutorial and ran into an issue during this step: fprime-util build aarch64-linux. I believe that this is meant to build for the x64 ARM system so that the pi can run it, but I am having an issue and this is what is outputted:

[100%] Linking CXX executable ../bin/aarch64-linux/GDep
  Toolchain 'aarch64-linux' should set CMAKE_SYSTEM_NAME to 'Generic' and set
  FPRIME_PLATFORM
Call Stack (most recent call first):
  fprime/cmake/FPrime.cmake:29 (include)
  CMakeLists.txt:13 (include)


/opt/toolchains/bin/../lib/gcc/aarch64-none-linux-gnu/10.2.1/../../../../aarch64-none-linux-gnu/bin/ld: /usr/lib/libwiringPi.so: error adding symbols: file in wrong format
collect2: error: ld returned 1 exit status
gmake[3]: *** [GDep/CMakeFiles/GDep.dir/build.make:217: bin/aarch64-linux/GDep] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:5288: GDep/CMakeFiles/GDep.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:5295: GDep/CMakeFiles/GDep.dir/rule] Error 2
gmake: *** [Makefile:1339: GDep] Error 2
[ERROR] CMake erred with return code 2
(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/GDep#


It gives me this warning at the top:

(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/GDep# fprime-util build aarch64-linux
-- [python3] python3 found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/python3
-- [fpp-tools] fpp-depend found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/fpp-depend
-- [fpp-tools] fpp-to-xml found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/fpp-to-xml
-- [fpp-tools] fpp-to-cpp found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/fpp-to-cpp
-- [fpp-tools] fpp-locate-defs found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/fpp-locate-defs
-- [FPRIME] Module locations: /mnt/c/Users/owenn/Command/fprime;/mnt/c/Users/owenn/Command
Linux/mnt/c/Users/owenn/Command/fprime/cmake/toolchain/aarch64-linux.cmake
aarch64-linux
CMake Warning at fprime/cmake/platform/platform.cmake:28 (message):
-- [FPRIME] Configuration module: /mnt/c/Users/owenn/Command/fprime/config
-- [FPRIME] Installation directory: /mnt/c/Users/owenn/Command/build-artifacts
-- Target build toolchain/platform: aarch64-linux/Linux
-- Including /mnt/c/Users/owenn/Command/fprime/cmake/platform/Linux.cmake
-- Requiring thread library


Any ideas?

**Reply by OwenNieuwenhuizen** [reply]
I tried using changing up the platform.cmake file to see if I could fix the problem, but I ended up reverting everything back to what it was there because it didn't seem to be working. Could my issue be another linker error or could it be something else?

**Comment by LeStarch** [comment]
/opt/toolchains/bin/../lib/gcc/aarch64-none-linux-gnu/10.2.1/../../../../aarch64-none-linux-gnu/bin/ld: /usr/lib/libwiringPi.so: error adding symbols: file in wrong format

It looks like it is linking against a local installation of wiringpi.  This will be the host architecture.  You'll need to figure out how to cross compile wiringpi and link to that

**Reply by LeStarch** [reply]
This discussion moved.  Marking and answer for part 1.
